### PR TITLE
DQ-302 Contract version 2.0

### PR DESCRIPTION
**Pull request by bichitra95**
## Change description

> This enhancement will change the behaviour of how the contract versioning is created. Now new version will not be created for every change to the contract, instead, it will keep on updating the same draft version and will create a new version incase the current version is published. 

## Type of change
- [ ] Bug fix (fixes an issue)
- [x] New feature (adds functionality)

## Related issues

> Fix [DQ-302](https://atlanhq.atlassian.net/browse/DQ-302) 

### Development

- [ ] Lint rules pass locally
- [x] Application changes have been tested thoroughly
- [ ] Automated tests covering modified code pass

### Security

- [ ] Security impact of change has been considered
- [ ] Code follows company security practices and guidelines

### Code review 

- [ ] Pull request has a descriptive title and context useful to a reviewer. Screenshots or screencasts are attached as necessary
- [ ] "Ready for review" label attached and reviewers assigned
- [ ] Changes have been reviewed by at least one other contributor
- [ ] Pull request linked to task tracker where applicable

[DQ-302]: https://atlanhq.atlassian.net/browse/DQ-302?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/contract/ContractPreProcessor.java [hunk 1]
```diff
@@ -79,14 +79,10 @@ private void processUpdateContract(AtlasEntity entity, EntityMutationContext con
         AtlasEntity existingContractEntity = entityRetriever.toAtlasEntity(vertex);
         // No update to relationships allowed for the existing contract version
         resetAllRelationshipAttributes(entity);
-        DataContract contract = DataContract.deserialize(contractString);
-        String existingContractString = getContractString(existingContractEntity);
-        boolean requestFromMigration = RequestContext.get().getRequestContextHeaders().getOrDefault(
-                "x-atlan-request-id", "").contains("json-to-yaml-migration");
-        if (!requestFromMigration && !StringUtils.isEmpty(contractString) &&
-                !contract.equals(DataContract.deserialize(existingContractString))) {
+        if (entity.getAttribute(ATTR_CERTIFICATE_STATUS) == DataContract.Status.VERIFIED.name() ||
```
Comment (nikhilbonte21) [inline repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/contract/ContractPreProcessor.java:82]: Why this check done on entity as well?

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/contract/ContractPreProcessor.java [hunk 2]
```diff
@@ -122,6 +118,9 @@ private void processCreateContract(AtlasEntity entity, EntityMutationContext con
 
         AtlasEntity currentVersionEntity = getCurrentVersion(associatedAsset.getEntity().getGuid());
         Long newVersionNumber = 1L;
+        if (currentVersionEntity == null && contract.getStatus() == DataContract.Status.VERIFIED) {
+            throw new AtlasBaseException("Can't create a new published version");
```
Comment (nikhilbonte21) [inline repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/contract/ContractPreProcessor.java:122]: I suggest using error code here, may be send OPERATION_NOT_SUPPORTED

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/contract/ContractPreProcessor.java [hunk 3]
```diff
@@ -130,13 +129,16 @@ private void processCreateContract(AtlasEntity entity, EntityMutationContext con
...
                 resetAllRelationshipAttributes(entity);
-                // No change in contract, metadata changed
+                // Contract is in draft state. Update the same version
                 updateExistingVersion(context, entity, currentVersionEntity);
                 newVersionNumber = currentVersionNumber;
             } else {
-                // contract changed (metadata might/not changed). Create new version.
+                // Current version is published. Creating a new draft version.
+                if (contract.getStatus() == DataContract.Status.VERIFIED) {
+                    throw new AtlasBaseException("Can't create a new published version");
```
Comment (nikhilbonte21) [inline repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/contract/ContractPreProcessor.java:140]: I suggest using error code here, may be send OPERATION_NOT_SUPPORTED

**Review by nikhilbonte21 — CHANGES_REQUESTED** [review]

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/contract/ContractPreProcessor.java (hunk 1 above)
Comment (bichitra95) [reply repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/contract/ContractPreProcessor.java:82]: If the entity state is already published, we shouldn't allow any edits to the entity.

**Review by nikhilbonte21 — APPROVED** [review]

**Review by sumandas0 — APPROVED** [review]

**Pull request merged**